### ISSUE TITLE
Ordering of analysis results (may work already -> needs testing)

### ISSUE THREAD

**Issue by aarppe** [UAlbertaALTLab/itwewina#18]
* either alphabetically or with descending lemma frequencyunder each POS.
* frequency as default, but some toggle to switch to alphabetical ordering.
* coded in XML source in rank=".." element as a number --> AA to add lemma frequencies to source code based on A-W corpus.

**Comment by aarppe** [comment]
I should add corpus-based lemma frequencies into the <rank>...</rank> fields in crk2eng and eng2crk dictionary source XML code. Testing needed only after that.

**Comment by aarppe** [comment]
Frequencies added to crkeng.xml, but whether the info is in the correct field for NDS to make use of that information needs to be checked.

**Comment by eddieantonio** [comment]
> Frequencies added to crkeng.xml, but whether the info is in the correct field for NDS to make use of that information needs to be checked.

I can also simply change the field, or even better, document which field has the weight applied to it, and in what format.

**Comment by eddieantonio** [comment]
From #6 (duplicate):

![img_20180705_115625](https://user-images.githubusercontent.com/2294397/42340882-7eceae6e-804e-11e8-81c1-4dc3ae0d96d9.jpg)

**Given** I executed a search,
**Then** the results should be listed in order of the most common lemma first.

**Comment by eddieantonio** [comment]
Another thought: I think we may want to have multi-tiered sorting by default. Not only does **lemma frequency** matter, but also **similarity to the query**. So I may want to incorporate edit distance into the ranking, but this may be more of a "nice-to-have" feature than something that is required right now.

Edit distance could be defined as Levenshtein distance, where differences in accents and aspiration only count for half the distance (because we expect people to drop accents and aspiration in the query).

**Comment by aarppe** [comment]
It would seem to make most sense that the ranking variable ('rank' in the current XML code) is attached to the language-lemma which is being ranked - currently, we've ranked the crk lemmas.

I'll provide an example of where I currently put this (the rank variable in the <l> field), based on my best understanding:

```
- <e src="nêhiyawêwin : itwêwina / Cree : Words">
-    <lg>
-       <l pos="N" rank="5988">acâhkos</l>
-       <lc>NA-1</lc>
-       <stem>acâhkos-</stem>
-    </lg>
-    <mg>
-    <tg xml:lang="eng">
-        <t pos="N">star</t>
-    </tg>
-    </mg>
- </e>
- 
```

Note that in the eng2crk direction (in engcrk.xml), the 'rank' variable is in a different field, the <mg> one:

```
- <e src="nêhiyawêwin : itwêwina / Cree : Words">
-    <lg>
-        <l pos="N">star</l>
-    </lg>
-    <mg rank="1">
-    <tg xml:lang="crk">
-       <re>star</re>
-       <re2>star</re2>
-       <t pos="N" type="AN">acâhkos</t>
-          <lc>=NA-1</lc>
-          <stem>acâhkos-</stem>
-          <key>star</key>
-          <key>little star</key>
-    </tg>
-    </mg>
-    <mg rank="2">
-    <tg xml:lang="crk">
-       <re>evening star</re>
-       <re2>evening star</re2>
-       <t pos="N" type="AN">otâkwanacâhkos</t>
-          <lc>=NA-1</lc>
-          <stem>otâkwanacâhkos-</stem>
-          <key>evening</key>
-          <key>star</key>
-    </tg>
-    </mg>
-    <tg xml:lang="crk">
-       <re>morning star</re>
-       <re2>morning star</re2>
-       <t pos="N" type="AN">wâpanacâhkos</t>
-          <lc>=NA-1</lc>
-          <stem>wâpanacâhkos-</stem>
-          <key>morning</key>
-          <key>star</key>
-    </tg>
-    </mg>
-    <mg rank="4">
-    <tg xml:lang="crk">
-       <re>comet, tailed star</re>
-       <re2>comet, tailed star</re2>
-       <t pos="I">acâhkos kâ-osôsit</t>
-          <lc>=INM</lc>
-          <stem></stem>
-          <key>comet</key>
-          <key>star</key>
-          <key>tail</key>
-    </tg>
-    </mg>
- ...
- </e>
```

**Comment by aarppe** [comment]
Yes, I agree on the longer term that we would much prefer a more sophisticated ranking procedure. Besides the lemma frequency, there's the tag, or better-yet tag-sequence frequency, or even better (in principle) the probability of the lemma and the associated tag sequence. And as you mention, distance between the search string and the normative form of the word.

Ranking the lemmas is something that can be done within NDS without mixing that with a weighted FST, so that's the cleanest feature to implement (right now), as the rest would require a standard weight to incorporate a weighted FST (that needs to be weighted every time it's recompiled).

At the same time, I'm sometimes thinking whether having only a few tiers of ranks, under which items are alphabetized, might be the best way to search things, or to what extent solely alphabetical sorting can have its merits. But these are considerations for later stages.

**Comment by eddieantonio** [comment]
Given the feedback we heard today, I think reranking the results should be given higher priority. The sorting of partial matches really leaves something to be desired.

**Comment by aarppe** [comment]
The partial matching was implemented because the old on-line Cree dictionary does that partial matching (in part because it doesn't have morphology), and a reason some people had heard for folks to prefer the old dictionary was that it gave more results, whereas our new one gave only exact ones. But lemma-frequency sorting would generally be the way to go, even though it's not always entirely reliable.

And I can create rankings for lemmas that do not have a corpus-frequency by using the weighted derivational analyzer, which gives an estimate of how likely the morpheme combination is (in Arok's lexical database), as a proxy for lemma frequency.
